File: 001/gradient_descent_parameter_fitting/gradient_descent_parameter_fitting.jl
avg = []
for i = 1:10
    start_time = time_ns()

    array =  [1 0 0;  1 1 1;  1 2 4]
    x = [1 0;  1 1;  1 2]
    y = [0;  1;  4]

    coeffs = Dict("beta_0" => 0.0, "beta_1" => 2.0)
    coeff_order = ["beta_0", "beta_1"]

    function regression_function(coeffs, inputs)
        return sum([coeffs[coeff] * term for (coeff, term) in zip(coeff_order, inputs)])
    end

    function rss(coeffs)
        return sum([(regression_function(coeffs, row[1:(length(y) - 1)]) - row[length(y)]) ^ 2 for row in eachrow(array)]) 
    end

    function calc_gradient(coeffs, learning_rate = 0.001)
        gradient = Dict()
        temp = Dict(beta => coeff for (beta, coeff) in coeffs)
        for key in coeff_order
            temp[key] += learning_rate
            plus_prediction = rss(temp)
            temp[key] -= 2 * learning_rate
            minus_prediction = rss(temp)
            temp[key] += learning_rate
            push!(gradient, key => (plus_prediction - minus_prediction) / (2 * learning_rate))
        end
        return gradient
    end

    function descend(coeffs, learning_rate = 0.001)
        gradient = calc_gradient(coeffs, learning_rate)
        for key in coeff_order
            coeffs[key] -= gradient[key] * learning_rate
        end
        return coeffs
    end

    for i = 1:10000
        coeffs = descend(coeffs)
    end

    append!(avg, time_ns() - start_time)
end

print("Julia: ", sum(avg) / 1000000000 / 10)

File: 001/gradient_descent_parameter_fitting/gradient_descent_parameter_fitting.py
import time

avg = []
for i in range(10):
    start_time = time.time()

    array =  [[1, 0, 0], [1, 1, 1],  [1, 2, 4]]
    x = [[0, 1], [1, 1], [2, 1]]
    y = [[0], [1], [4]]

    coeffs = {'beta_0': 0, 'beta_1': 2}
    coeff_order = ["beta_0", "beta_1"]

    def regression_function(coeffs, inputs):
        return sum([coeffs[coeff] * term for (coeff, term) in zip(coeff_order, inputs)])

    def rss(coeffs):
        return sum([(regression_function(coeffs, row[:len(y) - 1]) - row[len(y) - 1]) ** 2 for row in array]) 

    def calc_gradient(coeffs, learning_rate = 0.001):
        gradient = {}
        temp = {beta: coeff for beta, coeff in coeffs.items()}
        for beta in coeffs.keys():
            temp[beta] += learning_rate
            plus_prediction = rss(temp)
            temp[beta] -= 2 * learning_rate
            minus_prediction = rss(temp)
            temp[beta] += learning_rate
            gradient[beta] = (plus_prediction - minus_prediction) / (2 * learning_rate)
        return gradient

    def descend(coeffs, learning_rate = 0.001):
        new_coeffs = {beta: coeff for beta, coeff in coeffs.items()}
        gradient = calc_gradient(coeffs, learning_rate)
        for beta in coeffs.keys():
            new_coeffs[beta] -= gradient[beta] * learning_rate
        return new_coeffs

    for _ in range(10000):
        coeffs = descend(coeffs)

    avg.append(time.time() - start_time)

print('Python:', sum(avg) / 10)